### PR TITLE
Replace verbose azcore errors with more concise ones

### LINKED ISSUE #3778: Storage Account creation error prints the entire HTTP response
### What happened?

The following program:

```ts
const rg = new resources.ResourceGroup("rg");
const sa = new storage.StorageAccount("MySa", {
    resourceGroupName: rg.name,
    sku: {
        name: "Standard_LRS",
    },
    kind: "StorageV2",
});
```

fails to create a storage account, as expected. However, it prints out a very verbose error:

```
error: PUT https://management.azure.com/subscriptions/<redacted>/resourceGroups/rg-ce2m/providers/Microsoft.Storage/storageAccounts/MySa6198fc68
    --------------------------------------------------------------------------------
    RESPONSE 400: 400 Bad Request
    ERROR CODE: AccountNameInvalid
    --------------------------------------------------------------------------------
    {
      "error": {
        "code": "AccountNameInvalid",
        "message": "MySa6198fc68 is not a valid storage account name. Storage account name must be between 3 and 24 characters in length and use numbers and lower-case letters only."
      }
    }
    --------------------------------------------------------------------------------
```

This is likely a consequence of azcore migration. The earlier version would print

```
error: Code="AccountNameInvalid" Message="MySa6198fc68 is not a valid storage account name. Storage account name must be between 3 and 24 characters in length and use numbers and lower-case letters only."
```

### Example

See above

### Output of `pulumi about`

```
@pulumi/azure-native      2.76.0
@pulumi/pulumi            3.140.0
```

### Additional context

_No response_

### Contributing

Vote on this issue by adding a 👍 reaction. 
To contribute a fix for this issue, leave a comment (and link to your pull request, if you've opened one already).

### PR DESCRIPTION
**Pull request by thomas11**
Fixes #3778

Before:
```
$ pulumi up --skip-preview
...
Diagnostics:
  pulumi:pulumi:Stack (azn-http-errors-dev):
    error: update failed

  azure-native:storage:StorageAccount (sa):
    error: PUT https://management.azure.com/subscriptions/123/resourceGroups/resourceGroup84362436/providers/Microsoft.Storage/storageAccounts/INVALID
    --------------------------------------------------------------------------------
    RESPONSE 400: 400 Bad Request
    ERROR CODE: AccountNameInvalid
    --------------------------------------------------------------------------------
    {
      "error": {
        "code": "AccountNameInvalid",
        "message": "INVALID is not a valid storage account name. Storage account name must be between 3 and 24 characters in length and use numbers and lower-case letters only."
      }
    }
    --------------------------------------------------------------------------------

...
```

After:
```
$ PATH=$HOME/pulumi/pulumi-azure-native/bin:$PATH pulumi up --skip-preview
...
  azure-native:storage:StorageAccount (sa):
    error: Code="AccountNameInvalid" Message="INVALID is not a valid storage account name. Storage account name must be between 3 and 24 characters in length and use numbers and lower-case letters only."

...
```

### CODE REVIEW

**Review by danielrbradley — APPROVED** [review]

File: provider/pkg/azure/client_azcore.go [hunk 1]
```diff
@@ -517,3 +519,37 @@ func (r *requestAssertingTransporter) Do(req *http.Request) (*http.Response, err
 	r.assertions(r.t, req)
 	return nil, nil
 }
+
+func newResponseError(resp *http.Response) error {
```
Comment (mikhailshilkov) [inline provider/pkg/azure/client_azcore.go:523]: Do we want to call it in shouldRetryConflict too?

File: provider/pkg/azure/client_azcore.go [hunk 2]
```diff
@@ -517,3 +519,37 @@ func (r *requestAssertingTransporter) Do(req *http.Request) (*http.Response, err
...
+	}
+
+	body, err2 := runtime.Payload(resp)
+	if err2 != nil {
+		return err
+	}
+
+	var payload map[string]any
+	if err2 := json.Unmarshal(body, &payload); err2 != nil {
+		return err
```
Comment (mikhailshilkov) [inline provider/pkg/azure/client_azcore.go:537]: If body is not JSON (e.g. a string), do we really want to return a JSON deserialization error to the user? Shouldn't we show the string from the body?

File: provider/pkg/azure/client_azcore.go [hunk 3]
```diff
@@ -517,3 +519,37 @@ func (r *requestAssertingTransporter) Do(req *http.Request) (*http.Response, err
...
+	errMsg := string(body)
+	if e, ok := util.GetInnerMap(payload, "error"); ok {
+		if msg, ok := e["message"]; ok {
+			errMsg = msg.(string)
+		}
+	}
+
+	errCode := azcoreErr.ErrorCode
+	if errCode == "" {
+		errCode = fmt.Sprintf("%d", resp.StatusCode)
```
Comment (mikhailshilkov) [inline provider/pkg/azure/client_azcore.go:549]: Should we just return a message in this case? Or is Code= Message= format important for some reason?

Comment (mikhailshilkov) [reply provider/pkg/azure/client_azcore.go:549]: Maybe it's useful to return a status code

File: provider/pkg/azure/client_azcore.go (hunk 2 above)
Comment (mikhailshilkov) [reply provider/pkg/azure/client_azcore.go:537]: Ah, I misread it, it's `err` not `err2`

**Review by mikhailshilkov — APPROVED** [review]

File: provider/pkg/azure/client_azcore.go (hunk 1 above)
Comment (thomas11) [reply provider/pkg/azure/client_azcore.go:523]: The error in `shouldRetryConflict` is never shown to the user. We only access its `ErrorCode`.

**Comment by EronWright** [comment]
A possible follow-up is to handle the final error message of long-running operations, as returned by the poller as an `azcore.ResponseError`.

**Pull request merged**